### PR TITLE
Let Users pass dynamic path to read main and test apks

### PR DESCRIPTION
**Pull request by richaArora08**
https://browserstack.atlassian.net/browse/AASI-2063
This change will let user pass paths to their main and test apks.
Example command:-
./gradlew executeDebugTestsOnBrowserstack -PmainAPKPath=/Users/richa/Documents/gradle_richa/test/mainApk -PtestAPKPath=/Users/richa/Documents/gradle_richa/test/testApk
Latest APKs present in the given folder will be picked for execution.

### CODE REVIEW

File: test.rb [hunk 1]
```diff
@@ -59,9 +71,17 @@ def run_tests_args
...
   run_basic_espresso_test("./gradlew executeDebugTestsOnBrowserstack -PskipBuildingApks=true")
+  run_basic_espresso_test("./gradlew executeDebugTestsOnBrowserstack -PskipBuildingApks=false")
   print_separator
 end
 
+def run_tests_with_path_args
+  puts "\nRunning new test using ./gradlew with APK path args"
+  mainAPKPath = __dir__ + "/test/mainApk"
+  testAPKPAth = __dir__ + "/test/testApk"
+  run_espresso_test_with_path("./gradlew executeDebugTestsOnBrowserstack -PmainAPKPath=#{mainAPKPath} -PtestAPKPath=#{testAPKPAth}")
```
Comment (deepwired) [inline test.rb:82]: lets add a unit test for dynamic path + config file with instrumentation, cucumber options?

Comment (deepwired) [reply test.rb:82]: Lets also add a negative scenario where an apk is not present in the provided path

File: src/main/java/com/browserstack/gradle/BrowserStackTask.java [hunk 2]
```diff
@@ -158,12 +174,30 @@ public Map<String, Path> locateApks(boolean ignoreTestPath) throws IOException {
...
+    }
+    if(testAPKPath != null){
+      isAPKFileCreated[1] = true;
+      Files.find(Paths.get(testAPKPath),1, (filePath, fileAttr) -> isValidAPKFile(filePath, fileAttr))
+              .forEach(f -> {
+                testApkFiles.add(f);
+              });
+    }
+
     Files.find(Paths.get(dir), Constants.APP_SEARCH_MAX_DEPTH, (filePath, fileAttr) -> isValidFile(filePath, fileAttr))
```
Comment (deepwired) [inline src/main/java/com/browserstack/gradle/BrowserStackTask.java:193]: if both paths are provided no need to execute below loop on files

File: src/main/java/com/browserstack/gradle/BrowserStackTask.java [hunk 3]
```diff
@@ -158,12 +174,30 @@ public Map<String, Path> locateApks(boolean ignoreTestPath) throws IOException {
...
         .forEach(f -> {
 
           if (f.toString().endsWith("-androidTest.apk")) {
```
Comment (deepwired) [inline src/main/java/com/browserstack/gradle/BrowserStackTask.java:196]: might as well combine these if statements `!isAPKFileCreated[1] && f.toString().endsWith("-androidTest.apk")`

**Review by deepwired — CHANGES_REQUESTED** [review]

File: src/main/java/com/browserstack/gradle/BrowserStackTask.java (hunk 2 above)
Comment (richaArora08) [reply src/main/java/com/browserstack/gradle/BrowserStackTask.java:193]: Ok.

File: src/main/java/com/browserstack/gradle/BrowserStackTask.java (hunk 3 above)
Comment (richaArora08) [reply src/main/java/com/browserstack/gradle/BrowserStackTask.java:196]: Done

File: test.rb [hunk 4]
```diff
@@ -82,6 +94,14 @@ def run_tests_with_path_args
   run_espresso_test_with_path("./gradlew executeDebugTestsOnBrowserstack -PmainAPKPath=#{mainAPKPath} -PtestAPKPath=#{testAPKPAth}")
 end
 
+def run_test_with_incorrect_path
```
Comment (deepwired) [inline test.rb:97]: add variations, print statements
- when testapk path is provided
- when aut path is provided
- when both is provided

**Review by deepwired — CHANGES_REQUESTED** [review]

**Review by deepwired — DISMISSED** [review]
LGTM

File: test.rb [hunk 5]
```diff
@@ -34,6 +34,54 @@ def run_basic_espresso_test(gradle_command)
...
+    puts responses.join("\n")
+  end
+end
+
+def run_espresso_test_with_incorrect_path(gradle_command)
+  puts "Running #{gradle_command} with basic config and incorrect paths to both main and test apk"
+  stdout = run_command(gradle_command)
+  responses = stdout.lines.select{ |line| line.match(/DebugApp apk: null|TestApp apk: null/)}
+  if responses.count != 2
+    puts "✘ #{gradle_command} failed with error: #{responses}".red
```
Comment (deepwired) [inline test.rb:54]: if we are expecting this unit test to fail, can we change the color idication to green here, and red for the else condition?

Comment (deepwired) [reply test.rb:54]: resolved.

**Review by deepwired — DISMISSED** [review]

File: test.rb [hunk 6]
```diff
@@ -59,9 +142,149 @@ def run_tests_args
...
+   puts "\nRunning new test with relative correct main and non existing absolute test path"
+   run_espresso_test_with_one_existing_and_one_non_existing_apk_path("./gradlew executeDebugTestsOnBrowserstack -PmainAPKPath=#{$correct_relative_mainAPKPath} -PtestAPKPath=#{$incorrect_absolute_APKPath}", "test");
+   puts "\nRunning new test with relative correct main and non existing relative test path"
+   run_espresso_test_with_one_existing_and_one_non_existing_apk_path("./gradlew executeDebugTestsOnBrowserstack -PmainAPKPath=#{$correct_relative_mainAPKPath} -PtestAPKPath=#{$incorrect_relative_APKPath}", "test");
+end
+
+def run_test_with_ipa
+  puts "Running tests with ipa files"
+  mainAPKPath =  $current_path + "/test/mainApk/ipa"
+  testAPKPAth =  $current_path + "/test/mainApk/ipa"
```
Comment (sauravdas1997) [inline test.rb:277]: This will pick the same app, can we point it to any other path with apk test

Comment (richaArora08) [reply test.rb]: Changes pushed.

**Review by sauravdas1997 — APPROVED** [review]

**Review by deepwired — APPROVED** [review]
LGTM

**Pull request merged**